### ISSUE TITLE
Inductor doesn't fuse outer dimension softmax into a single kernel.

### ISSUE THREAD

**Issue by Chillee** [pytorch/pytorch#93718]

**Comment by ezyang** [comment]
@Chillee is this fixed now?

**Comment by ydwu4** [comment]
Repro:
```python
import torch
import time

def bench_inductor(f):
    f = torch.compile(f)
    for _ in range(2):
        out = f()
    torch.cuda.synchronize()
    iters = 100
    begin = time.time()
    for _ in range(iters):
        f()
    torch.cuda.synchronize()
    print(f"{(time.time()-begin)*1e6/iters},")

def softmax_torch(x):
    dim = 0
    x_max = torch.amax(x, dim, keepdim=True)
    unnormalized = torch.exp(x * x_max)
    result = unnormalized / torch.sum(unnormalized, dim, keepdim=True)
    return result

i = 5
softmax_inputs = [(2**i, 2**(26-i))]
shapes = softmax_inputs
device = 'cuda'
inps = [torch.ones(shape, device=device) if isinstance(shape, tuple) else torch.tensor(shape(), device=device) for shape in shapes]
bench_inductor(lambda: softmax_torch(*inps))
```
It runs successfully and inductor generates following  code:
```python

from ctypes import c_void_p, c_long
import torch
import math
import random
import os
import tempfile
from math import inf, nan
from torch._inductor.hooks import run_intermediate_hooks
from torch._inductor.utils import maybe_profile
from torch._inductor.codegen.memory_planning import _align as align

from torch import device, empty, empty_strided
from torch._inductor.codecache import AsyncCompile
from torch._inductor.select_algorithm import extern_kernels

aten = torch.ops.aten
inductor_ops = torch.ops.inductor
assert_size_stride = torch._C._dynamo.guards.assert_size_stride
alloc_from_pool = torch.ops.inductor._alloc_from_pool
reinterpret_tensor = torch.ops.inductor._reinterpret_tensor
async_compile = AsyncCompile()


# kernel path: /tmp/torchinductor_yidi/ze/cze4f423y5dhpyykn3sawteup7kresz2j2qvrxeyauqymcadpp6z.py
# Source Nodes: [mul, sum_1, unnormalized, x_max], Original ATen: [aten.amax, aten.exp, aten.mul, aten.sum]
# mul => mul
# sum_1 => sum_1
# unnormalized => exp
# x_max => amax
triton_per_fused_amax_exp_mul_sum_0 = async_compile.triton('triton_', '''
import triton
import triton.language as tl
from torch._inductor.ir import ReductionHint
from torch._inductor.ir import TileHint
from torch._inductor.triton_heuristics import AutotuneHint, persistent_reduction
from torch._inductor.utils import instance_descriptor
from torch._inductor import triton_helpers

@persistent_reduction(
    size_hints=[2097152, 32],
    reduction_hint=ReductionHint.DEFAULT,
    filename=__file__,
    triton_meta={'signature': {0: '*fp32', 1: '*fp32', 2: '*fp32', 3: 'i32', 4: 'i32'}, 'device': 0, 'device_type': 'cuda', 'constants': {}, 'configs': [instance_descriptor(divisible_by_16=(0, 1, 2, 3, 4), equal_to_1=(), ids_of_folded_args=(), divisible_by_8=(3, 4))]},
    inductor_meta={'autotune_hints': set(), 'kernel_name': 'triton_per_fused_amax_exp_mul_sum_0', 'mutated_arg_names': []}
)
@triton.jit
def triton_(in_ptr0, out_ptr0, out_ptr1, xnumel, rnumel, XBLOCK : tl.constexpr):
    xnumel = 2097152
    rnumel = 32
    RBLOCK: tl.constexpr = 32
    xoffset = tl.program_id(0) * XBLOCK
    xindex = xoffset + tl.arange(0, XBLOCK)[:, None]
    xmask = xindex < xnumel
    rindex = tl.arange(0, RBLOCK)[None, :]
    rmask = rindex < rnumel
    r1 = rindex
    x0 = xindex
    tmp0 = tl.load(in_ptr0 + (x0 + (2097152*r1)), rmask, other=0.0)
    tmp1 = tl.broadcast_to(tmp0, [XBLOCK, RBLOCK])
    tmp3 = tl.where(rmask, tmp1, float("-inf"))
    tmp4 = triton_helpers.max2(tmp3, 1)[:, None]
    tmp5 = tmp0 * tmp4
    tmp6 = tl.exp(tmp5)
    tmp7 = tl.broadcast_to(tmp6, [XBLOCK, RBLOCK])
    tmp9 = tl.where(rmask, tmp7, 0)
    tmp10 = tl.sum(tmp9, 1)[:, None]
    tl.store(out_ptr0 + (x0), tmp4, None)
    tl.store(out_ptr1 + (x0), tmp10, None)
''')

import triton
import triton.language as tl
from torch._inductor.triton_heuristics import grid, start_graph, end_graph
from torch._C import _cuda_getCurrentRawStream as get_cuda_stream


# kernel path: /tmp/torchinductor_yidi/s2/cs2asyrgh4mbe2md2jtmgqiovyivgjnhxhjxkabhso35k6opspbk.py
# Source Nodes: [mul, result, unnormalized], Original ATen: [aten.div, aten.exp, aten.mul]
# mul => mul
# result => div
# unnormalized => exp
triton_poi_fused_div_exp_mul_1 = async_compile.triton('triton_', '''
import triton
import triton.language as tl
from torch._inductor.ir import ReductionHint
from torch._inductor.ir import TileHint
from torch._inductor.triton_heuristics import AutotuneHint, pointwise
from torch._inductor.utils import instance_descriptor
from torch._inductor import triton_helpers

@pointwise(
    size_hints=[67108864], 
    filename=__file__,
    triton_meta={'signature': {0: '*fp32', 1: '*fp32', 2: '*fp32', 3: '*fp32', 4: 'i32'}, 'device': 0, 'device_type': 'cuda', 'constants': {}, 'configs': [instance_descriptor(divisible_by_16=(0, 1, 2, 3, 4), equal_to_1=(), ids_of_folded_args=(), divisible_by_8=(4,))]},
    inductor_meta={'autotune_hints': set(), 'kernel_name': 'triton_poi_fused_div_exp_mul_1', 'mutated_arg_names': []},
    min_elem_per_thread=0
)
@triton.jit
def triton_(in_ptr0, in_ptr1, in_ptr2, out_ptr0, xnumel, XBLOCK : tl.constexpr):
    xnumel = 67108864
    xoffset = tl.program_id(0) * XBLOCK
    xindex = xoffset + tl.arange(0, XBLOCK)[:]
    xmask = xindex < xnumel
    x2 = xindex
    x0 = xindex % 2097152
    tmp0 = tl.load(in_ptr0 + (x2), None)
    tmp1 = tl.load(in_ptr1 + (x0), None, eviction_policy='evict_last')
    tmp4 = tl.load(in_ptr2 + (x0), None, eviction_policy='evict_last')
    tmp2 = tmp0 * tmp1
    tmp3 = tl.exp(tmp2)
    tmp5 = tmp3 / tmp4
    tl.store(out_ptr0 + (x2), tmp5, None)
''')


async_compile.wait(globals())
del async_compile

def call(args):
    arg0_1, = args
    args.clear()
    assert_size_stride(arg0_1, (32, 2097152), (2097152, 1))
    with torch.cuda._DeviceGuard(0):
        torch.cuda.set_device(0) # no-op to ensure context
        buf0 = empty((1, 2097152), device='cuda', dtype=torch.float32)
        buf1 = empty((1, 2097152), device='cuda', dtype=torch.float32)
        # Source Nodes: [mul, sum_1, unnormalized, x_max], Original ATen: [aten.amax, aten.exp, aten.mul, aten.sum]
        stream0 = get_cuda_stream(0)
        triton_per_fused_amax_exp_mul_sum_0.run(arg0_1, buf0, buf1, 2097152, 32, grid=grid(2097152), stream=stream0)
        buf2 = empty((32, 2097152), device='cuda', dtype=torch.float32)
        # Source Nodes: [mul, result, unnormalized], Original ATen: [aten.div, aten.exp, aten.mul]
        triton_poi_fused_div_exp_mul_1.run(arg0_1, buf0, buf1, buf2, 67108864, grid=grid(67108864), stream=stream0)
        del arg0_1
        return (buf2, )


def benchmark_compiled_module(times=10, repeat=10):
    from torch._dynamo.testing import rand_strided
    from torch._inductor.utils import print_performance
    arg0_1 = rand_strided((32, 2097152), (2097152, 1), device='cuda:0', dtype=torch.float32)
    return print_performance(lambda: call([arg0_1]), times=times, repeat=repeat)


if __name__ == "__main__":
    from torch._inductor.wrapper_benchmark import compiled_module_main
    compiled_module_main('None', benchmark_compiled_module)
```
